cache/cache.go

```diff
@@ -0,0 +1,107 @@
+package cache
+
+import (
+	"container/list"
+)
+
+// Node represents a node eligible for caching.
+type Node interface {
+	GetKey() []byte
+}
+
+// Cache is an in-memory structure to persist nodes for quick access.
+// Please see lruCache for more details about why we need a custom
+// cache implementation.
+type Cache interface {
+	// Adds node to cache. If full and had to remove the oldest element,
+	// returns the oldest, otherwise nil.
+	Add(node Node) Node
```

Comment:
There should never be a nil Node added. Would it be sufficient to add a comment saying that the contract for this method is that `node` is never nil?

From reading [this discussion](https://stackoverflow.com/questions/13476349/check-for-nil-and-nil-interface-in-go), I could either:
- use reflection to check against nil in the `Add` method
- define `IsNil()` method for the `Node` interface and implement it for every node struct.

However, if we were to go with one of the approaches, I think that this would be trading performance for safety unnecessarily at such a low database layer.